File: src/Samples/Interop_Samples/HtmlPresenter/NumericTextBox.vb
Imports System.Windows
Imports CSHTML5.Internal
Imports CSHTML5.Native.Html.Controls
Imports OpenSilver

Namespace TestNumericTextBox

    Public Class NumericTextBox
        Inherits HtmlPresenter

        Private _value As Integer = 0
        Private _domElement As Object

        Public Sub New()
            Html = "<input type='number' pattern='[0-9]*' style='width:100%;height:100%'>"
            AddHandler Loaded, AddressOf NumericTextBox_Loaded
        End Sub

        Public Property Value As Integer
            Get
                If _domElement IsNot Nothing Then
                    Dim id As String = TryCast(_domElement, INTERNAL_HtmlDomElementReference).UniqueIdentifier
                    Dim valueString As String = Interop.ExecuteJavaScriptGetResult(Of String)($"{id}.firstChild.firstChild.value")
                    Dim valueInt As Integer
                    If Integer.TryParse(valueString, valueInt) Then
                        _value = valueInt
                    End If
                End If
                Return _value
            End Get
            Set(ByVal value As Integer)
                _value = value
                If _domElement IsNot Nothing Then
                    UpdateValue()
                End If
            End Set
        End Property

        Private Sub NumericTextBox_Loaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
            _domElement = Interop.GetDiv(Me)
            UpdateValue()
        End Sub

        Private Sub UpdateValue()
            Interop.ExecuteJavaScriptVoidAsync("$0.firstChild.firstChild.value = $1", _domElement, _value)
        End Sub
    End Class

End Namespace

File: src/OpenSilver.Showcase/Samples/Interop_Samples/HtmlPresenter/NumericTextBox.fs
namespace TestNumericTextBox

open System
open CSHTML5.Internal
open CSHTML5.Native.Html.Controls
open OpenSilver

type NumericTextBox() as this =
    inherit HtmlPresenter()

    let mutable value = 0
    let mutable domElement: obj = null

    do
        this.Html <- "<input type='number' pattern='[0-9]*' style='width:100%;height:100%'>"
        this.Loaded.Add(fun _ -> this.NumericTextBox_Loaded())

    member this.Value
        with get() =
            if not (isNull domElement) then
                let id = (domElement :?> INTERNAL_HtmlDomElementReference).UniqueIdentifier
                let valueStr = Interop.ExecuteJavaScriptGetResult<string>($"{id}.firstChild.firstChild.value")
                match Int32.TryParse(valueStr) with
                | (true, parsedValue) -> value <- parsedValue
                | _ -> ()
            value
        and set(v) =
            value <- v
            if not (isNull domElement) then
                this.UpdateValue()

    member private this.NumericTextBox_Loaded() =
        domElement <- Interop.GetDiv(this)
        this.UpdateValue()

    member private this.UpdateValue() =
        Interop.ExecuteJavaScriptVoidAsync("$0.firstChild.firstChild.value = $1", domElement, value)
